src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -0,0 +1,527 @@
+            var seen = new HashSet<DependencyNode>();
+            var frameworkReducer = new FrameworkReducer();
+            var resultWarningProperties = new PackageSpecificWarningProperties();
+            var packageNoWarn = new Dictionary<string, ISet<NuGetLogCode>>(StringComparer.OrdinalIgnoreCase);
+
+            // All the packages in parent project's closure. 
+            // Once we have collected data for all of these, we can exit.
+            var parentPackageDependencies = new HashSet<string>(
+                targetGraph.Flattened.Where(d => d.Key.Type == LibraryType.Package).Select(d => d.Key.Name));
+
+            var parentTargetFramework = targetGraph.Framework;
+
+            // Add all dependencies into a dict for a quick transitive lookup
+            foreach (var dependencyGraphItem in targetGraph.Flattened.OrderBy(d => d.Key.Name))
+            {
+                WarningPropertiesCollection nodeWarningProperties = null;
+
+                if (IsProject(dependencyGraphItem.Key.Type))
+                {
+                    var localMatch = dependencyGraphItem.Data.Match as LocalMatch;
+                    var nodeProjectSpec = GetNodePackageSpec(localMatch);
+                    var nearestFramework = frameworkReducer.GetNearest(
+                        parentTargetFramework, 
+                        nodeProjectSpec.TargetFrameworks.Select(tfi => tfi.FrameworkName));
+
+                    // Get the WarningPropertiesCollection from the PackageSpec
+                    nodeWarningProperties = GetNodeWarningProperties(nodeProjectSpec, nearestFramework);
+                }
+
+                var lookUpNode = new LookUpNode()
+                {
+                    Dependencies = dependencyGraphItem.Data.Dependencies,
+                    WarningPropertiesCollection = nodeWarningProperties
+                };
+
+                dependencyMapping[dependencyGraphItem.Key.Name] = lookUpNode;
+            }
+
+            // Get the direct dependencies for the parent project to seed the queue
+            var parentDependencies = dependencyMapping[parentProjectName];
+
+            // Seed the queue with the parent project's direct dependencies
+            AddDependenciesToQueue(parentDependencies.Dependencies,
+                queue,
+                seen,
+                parentWarningPropertiesCollection);
+
+            // Add the parent project to the seen set to prevent adding it back to the queue
+            seen.Add(new DependencyNode(id: parentProjectName,
+                isProject: true,
+                warningPropertiesCollection: parentWarningPropertiesCollection));
+
+            // start taking one node from the queue and get all of it's dependencies
+            while (queue.Count > 0)
+            {
+                var node = queue.Dequeue();
+                if (!seen.Contains(node))
+                {
+                    // Add the node to seen set
+                    seen.Add(node);
```

Comment:
Use seen.Add(node) above, it returns true if seen.Contains was false, this code is currently doing the look up twice